pmacFilterControlWrapper.py

```diff
@@ -0,0 +1,366 @@
+
+                out_key = f"filter_set_{i}_out_pos_{j}"
+                out_value = builder.aOut(
+                    f"FILTER_SET:{i}:OUT:{j}",
+                    initial_value=0,
+                    on_update=lambda _, i=i: self._set_out_pos(i),
+                )
+                self.filter_sets_out[filter_set_key][out_key] = out_value
+
+    async def run_forever(self) -> None:
+
+        print("Connecting to ZMQ stream...")
+
+        req_status = b'{"command":"status"}'
+
+        self._send_message(req_status)
+        # resp = await self.zmq_stream.get_response()
+
+        # if resp:
+        #     print("Connected.")
+
+        await asyncio.gather(
+            *[
+                self.monitor_responses(self.zmq_stream),
+                self.monitor_responses(self.event_stream),
+                self.zmq_stream.run_forever(),
+                self.event_stream.run_forever(),
+                self._query_status(),
+            ]
+        )
+
+    async def monitor_responses(self, zmq_stream: ZeroMQAdapter) -> None:
+        running = True
+        while running:
+            resp = await zmq_stream.get_response()
+            resp_json = json.loads(resp[0])
+
+            if "status" in resp_json:
+                status = resp_json["status"]
+
+                self._handle_status(status)
+
+            running = zmq_stream.check_if_running()
+
+    async def _query_status(self) -> None:
+        running = True
+        while running:
+            if self.zmq_stream.running:
+                req_status = b'{"command":"status"}'
+                self._send_message(req_status)
+                running = self.zmq_stream.check_if_running()
+            await asyncio.sleep(0.1)
+
+    def _handle_status(self, status: json) -> None:
+
+        state = status["state"]
+        self.state.set(state)
+
+        version = status["version"]
+        self.version.set(str(version))
+
+        process_duration = status["process_duration"]
+        self.process_duration.set(process_duration)
+
+        process_period = status["process_period"]
+        self.process_period.set(process_period)
+
+        last_received_frame = status["last_received_frame"]
+        self.last_frame_received.set(last_received_frame)
+
+        last_processed_frame = status["last_processed_frame"]
+        self.last_frame_processed.set(last_processed_frame)
+
+        time_since_last_frame = status["time_since_last_message"]
+        self.time_since_last_frame.set(time_since_last_frame)
+
+        current_attenuation = status["current_attenuation"]
+        self.current_attenuation.set(current_attenuation)
+
+    def _send_message(self, message: bytes) -> bytes:
+        self.zmq_stream.send_message([message])
+
+    def _set_mode(self, mode: int) -> None:
+
+        # Set mode for PFC
+        mode_config = json.dumps({"command": "configure", "params": {"mode": mode}})
+        self._send_message(codecs.encode(mode_config, "utf-8"))
```

Comment:
You could make a `_send_command` method that just takes the `"params"` dict so that you don't have to keep repeating this.